src/lando/main/scm/git.py

```diff
@@ -0,0 +1,258 @@
+
+        This method uses the Git commands to update the repository
+        located at the given pull path to the specified target changeset.
+
+        Args:
+            pull_path (str): The path to pull from.
+            target_cset (str): The target changeset to update the repository to.
+
+        Returns:
+            str: The target changeset
+        """
+        branch = target_cset or self.default_branch
+        self.clean_repo()
+        self._git_run("pull", "--prune", pull_path)
+        self._git_run("checkout", "--force", "-B", branch)
+        return self.head_ref()
+
+    def clean_repo(self, *, strip_non_public_commits: bool = True):
+        """Reset the local repository to the origin"""
+        if strip_non_public_commits:
+            self._git_run(
+                "reset", "--hard", f"origin/{self.default_branch}", cwd=self.path
+            )
+        self._git_run("clean", "-fdx", cwd=self.path)
+
+    def format_stack_amend(self) -> Optional[list[str]]:
+        """Amend the top commit in the patch stack with changes from formatting."""
+        self._git_run("commit", "--all", "--amend", "--no-edit")
+        return [self.get_current_node()]
+
+    def format_stack_tip(self, commit_message: str) -> Optional[list[str]]:
+        """Add an autoformat commit to the top of the patch stack.
+
+        Return the commit hash of the autoformat commit as a `str`,
+        or return `None` if autoformatting made no changes.
+        """
+        self._git_run("commit", "--all", "--message", commit_message)
+        return [self.get_current_node()]
+
+    def get_current_node(self) -> str:
+        """Return the commit_id of the tip of the current branch"""
+        return self._git_run("rev-parse", "HEAD")
+
+    @property
+    def repo_is_initialized(self) -> bool:
+        """Determine whether the target repository is initialised"""
+        if not Path(self.path).exists():
+            return False
+
+        try:
+            self._git_run("status", cwd=self.path)
+        except SCMException:
+            return False
+
+        return True
+
+    @classmethod
+    def repo_is_supported(cls, path: str) -> bool:
+        """Determine wether the target repository is supported by this concrete implementation"""
+        try:
+            cls._git_run("ls-remote", path)
+        except SCMException:
+            return False
+
+        return True
+
+    @classmethod
+    def _git_run(
+        cls, *args, cwd: Optional[str] = None, must_succeed: bool = True
+    ) -> str:
+        """Run a git command and return full output.
+
+        Parameters:
+
+        args: list[str]
+            Arguments to git
+
+        cwd: str
+            Optional path to work in, default to '/'
+
+        Returns:
+            str: the standard output of the command
+        """
+        path = cwd or "/"
+        command = ["git"] + list(args)
+        logger.debug("Running " + " ".join(command) + " in " + path)
```

Comment:
nit: also this may be more readable as an f-string.

Reply:
I took a different approach, and used the `extra` parameter, to more closely match the logging from hg.